dace/codegen/targets/fpga.py

```diff
@@ -160,68 +163,242 @@ def on_target_used(self) -> None:
+                                              ScopeSubgraphView],
+                           dependencies: dict):
+        '''
+            Finds subgraphs of an SDFGState or ScopeSubgraphView that correspond to kernels.
+            This is done by looking at the kernel ID associated to each node ('_kernel' field).
+            :param graph, the state/subgraph to consider
+            :param dependencies: a dictionary containing for each kernel ID, the IDs of the kernels from which it
+                depends on
+            :return a list of tuples (subgraph, kernel ID) topologically ordered according kernel dependencies.
+        '''
+        from dace.sdfg.scope import ScopeSubgraphView
+
+        if not isinstance(graph, (dace.sdfg.SDFGState, ScopeSubgraphView)):
+            raise TypeError(
+                "Expected SDFGState or ScopeSubgraphView, got: {}".format(
+                    type(graph).__name__))
+
+        subgraphs = collections.defaultdict(
+            set)  # {kernel_id: {nodes in subgraph}}
+
+        # Go over the nodes and populate the kernels subgraphs
+        for node, node_state in graph.all_nodes_recursive():
+            if hasattr(node, "_kernel"):
+                if node._kernel in subgraphs:
+                    subgraphs[node._kernel].add(node)
+                else:
+                    subgraphs[node._kernel] = {node}
+
+            # add this node to the corresponding subgraph
+            if isinstance(node, dace.nodes.AccessNode):
+                # AccessNodes can be read from multiple kernels, so
+                # check all out edges
+
+                start_nodes = [e.dst for e in node_state.out_edges(node)]
+                for n in start_nodes:
+                    if hasattr(n, "_kernel"):
+                        if n._kernel in subgraphs:
+                            subgraphs[n._kernel].add(node)
+                        else:
+                            subgraphs[n._kernel] = {node}
+
+        # Now stick each of the found components together in a ScopeSubgraphView and return
+        # them. Sort according kernel dependencies order.
+
+        # Build a dependency graph
+        import networkx as nx
+        kernels_graph = nx.DiGraph()
+        for k in subgraphs.keys():
+            # we could have no dependencies at all
+            kernels_graph.add_node(k)
+            if k in dependencies:
+                kernel_dependencies = dependencies[k]
+                for p in kernel_dependencies:
+                    kernels_graph.add_edge(p, k)
+
+        subgraph_views = []
+        all_nodes = graph.nodes()
+
+        # Use topological sort to order kernels according to their dependencies
+        for kernel_id in nx.topological_sort(kernels_graph):
+            # Return the subgraph and the kernel id
+            subgraph_views.append((ScopeSubgraphView(
+                graph, [n for n in all_nodes if n in subgraphs[kernel_id]],
```

Comment:
Can't use use `subgraphs[kernel_id]` directly? Or do you need the ordering? (You might also be able to make this set a list instead, see other comment)

Reply:
No, I need the ordering: when I spawn a kernel, all the predecessors must be already spawned. Kernel_id numbering does not reflect this

Reply:
I don't like having to build another data structure for this, but I don't have an immediate, cleaner solution

Reply:
Why not make `subgraphs` a dictionary to lists instead of sets?

Reply:
How that can help here?

Reply:
How about flipping it, make `all_nodes = set(graph.nodes())`, and then do `n for n in subgraphs[kernel_id] if n in all_nodes` if subgraphs is a list, or if both are sets, just do `all_nodes & subgraphs[kernel_id]`

Reply:
Now I get. We were talking about two different things.